src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,1054 @@
+  }
+}
+
+template <typename R>
+std::shared_ptr<ExpressionResult<R>> ExpressionEvaluator::_evaluate_in_expression(const InExpression& expression) {
+  Fail("InExpression supports only int32_t as result");
+}
+
+template <>
+std::shared_ptr<ExpressionResult<int32_t>> ExpressionEvaluator::_evaluate_predicate_expression<int32_t>(
+    const AbstractPredicateExpression& predicate_expression) {
+  switch (predicate_expression.predicate_condition) {
+    case PredicateCondition::Equals:
+    case PredicateCondition::LessThanEquals:
+    case PredicateCondition::GreaterThanEquals:
+    case PredicateCondition::GreaterThan:
+    case PredicateCondition::NotEquals:
+    case PredicateCondition::LessThan:
+      return _evaluate_binary_predicate_expression<int32_t>(
+          static_cast<const BinaryPredicateExpression&>(predicate_expression));
+
+    case PredicateCondition::Between: {
+      // `a BETWEEN b AND c` is evaluated by transforming it to `a >= b AND a <= c` instead of evaluating it with a
+      // dedicated algorithm. This is because three expression data types (from three arguments) generate many type
+      // combinations and thus lengthen compile time and increase binary size notably.
+
+      const auto& between_expression = static_cast<const BetweenExpression&>(predicate_expression);
+      const auto gte_expression = greater_than_equals(between_expression.value(), between_expression.lower_bound());
+      const auto lte_expression = less_than_equals(between_expression.value(), between_expression.upper_bound());
+
+      const auto gte_lte_expression = and_(gte_expression, lte_expression);
+
+      return evaluate_expression_to_result<int32_t>(*gte_lte_expression);
+    }
+
+    case PredicateCondition::In:
+      return _evaluate_in_expression<int32_t>(static_cast<const InExpression&>(predicate_expression));
+
+    case PredicateCondition::Like:
+    case PredicateCondition::NotLike:
+      return _evaluate_like_expression<int32_t>(static_cast<const BinaryPredicateExpression&>(predicate_expression));
+
+    case PredicateCondition::IsNull:
+    case PredicateCondition::IsNotNull:
+      return _evaluate_is_null_expression<int32_t>(static_cast<const IsNullExpression&>(predicate_expression));
+  }
+}
+
+template <typename R>
+std::shared_ptr<ExpressionResult<R>> ExpressionEvaluator::_evaluate_predicate_expression(
+    const AbstractPredicateExpression& expression) {
+  Fail("Can only evaluate predicates to int32_t (aka bool)");
+}
+
+template <typename R>
+std::shared_ptr<ExpressionResult<R>> ExpressionEvaluator::_evaluate_column_expression(
+    const PQPColumnExpression& column_expression) {
+  Assert(_chunk, "Cannot access Columns in this Expression as it doesn't operate on a Table/Chunk");
+
+  const auto& column = *_chunk->get_column(column_expression.column_id);
+  Assert(column.data_type() == data_type_from_type<R>(), "Can't evaluate column to different type");
+
+  _ensure_column_materialization(column_expression.column_id);
```

Comment:
"ensure" can mean two things: Fail if it is not materialized or materialize if non yet done.

Reply:
I really couldn't come up with a better name, though I tried. Suggestions? `_materialize_column_if_not_yet_materialized()` sounds stupidly verbose, but may actually be better?

Reply:
I could live with that.